### PR TITLE
Alc5658 driver changes

### PR DESCRIPTION
**Pull request by btheosam**
Alc5658 driver don't register for underrun i2s error notification

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check OK.

File: os/drivers/audio/alc5658.c [hunk 1]
```diff
@@ -713,6 +713,25 @@ static int alc5658_shutdown(FAR struct audio_lowerhalf_s *dev)
...
+ ****************************************************************************/
+static void alc5658_io_err_cb(FAR struct i2s_dev_s *dev, FAR void *arg, int flags)
+{
+	FAR struct alc5658_dev_s *priv = (FAR struct alc5658_dev_s *)arg;
+
+	/* Call upper callback, let it post msg to user q
+	 * apb is set NULL, okay? Rethink
+	*/
+	auddbg("Received error with flags: 0x%x\n", flags);
+	priv->dev.upper(priv->dev.priv, AUDIO_CALLBACK_IOERR, NULL, flags);
```
Comment (kishore-sn) [inline os/drivers/audio/alc5658.c]: I think the flags must not be passed directly to upper layer. The upper layers cannot make sense of the flags related to I2S. The codec is responsible for converting the flags into some form that the upper layers can understand. 
My suggestion is to add new error codes in audio.h file for overflow and underflow scenario. This function has to check the I2S error and send appropriate error codes to upper layer. 

Also, need to add code in audio.c for AUDIO_CALLBACK_IOERR.

File: os/drivers/audio/alc5658.c [hunk 2]
```diff
@@ -729,41 +748,34 @@ static int alc5658_start(FAR struct audio_lowerhalf_s *dev)
...
-	if (entry == -1) {
-		return -EINVAL;
+	if (priv->running) {
+		goto alcstart_ok;
 	}
 
-	alc5658_exec_i2c_script(priv, g_sample_entry[entry].script, g_sample_entry[entry].size / sizeof(t_codec_init_script_entry));
-	alc5658_exec_i2c_script(priv, codec_init_inout_script2, sizeof(codec_init_inout_script2) / sizeof(t_codec_init_script_entry));
+	/* Register cb for io error */
+	I2S_ERR_CB_REG(priv->i2s, alc5658_io_err_cb, priv);
```
Comment (kishore-sn) [inline os/drivers/audio/alc5658.c]: Other than overflow and underflow, there can be a i2s timeout error also. I think that also needs to be handled.

File: os/drivers/audio/alc5658.c [hunk 3]
```diff
@@ -912,7 +922,7 @@ static int alc5658_enqueuebuffer(FAR struct audio_lowerhalf_s *dev, FAR struct a
 
 	if (priv->inout) {		/* record */
 		ret = I2S_RECEIVE(priv->i2s, apb, alc5658_rxtxcallback, priv, CONFIG_ALC5658_I2S_TIMEOUT);
-	} else {					/* playback */
+	} else {				/* playback */
```
Comment (junmin-kim) [inline os/drivers/audio/alc5658.c]: Is this diff necessary and related with commit message?
It seems better to remove unnecessary diff chunck.

File: os/drivers/audio/alc5658.c (hunk 1 above)
Comment (Taejun-Kwon) [reply os/drivers/audio/alc5658.c]: I have a same opinion as kishore's

File: os/drivers/audio/alc5658.c [hunk 4]
```diff
@@ -729,41 +748,34 @@ static int alc5658_start(FAR struct audio_lowerhalf_s *dev)
...
-	int entry;
-	if (priv->running) {
-		return OK;
-	}
-
 	audvdbg(" alc5658_start Entry\n");
-	alc5658_exec_i2c_script(priv, codec_init_inout_script1, sizeof(codec_init_inout_script1) / sizeof(t_codec_init_script_entry));
+	alc5658_takesem(&priv->devsem);
 
-	entry = alc5658_get_sample_rate_script(priv->samprate);
```
Comment (Taejun-Kwon) [inline os/drivers/audio/alc5658.c]: If we remove code here, can we change sample rate??

File: os/drivers/audio/alc5658.c [hunk 5]
```diff
@@ -729,41 +748,34 @@ static int alc5658_start(FAR struct audio_lowerhalf_s *dev)
...
-	if (entry == -1) {
-		return -EINVAL;
+	if (priv->running) {
+		goto alcstart_ok;
```
Comment (Taejun-Kwon) [inline os/drivers/audio/alc5658.c]: goto end_with_sem; or start_with_sem; or you can rename xxxx_with_sem.

File: os/drivers/audio/alc5658.c [hunk 6]
```diff
@@ -985,10 +986,8 @@ static int alc5658_ioctl(FAR struct audio_lowerhalf_s *dev, int cmd, unsigned lo
 #endif
 	}
 	break;
-
-
 	default:
-		audvdbg("Ignored\n");
+		auddbg("alc5658_ioctl received unkown cmd 0x%x\n", cmd);
```
Comment (Taejun-Kwon) [inline os/drivers/audio/alc5658.c]: good change for information. but verbose is correct, I think.

File: os/drivers/audio/alc5658.c [hunk 7]
```diff
@@ -1274,28 +1265,25 @@ FAR struct audio_lowerhalf_s *alc5658_initialize(FAR struct i2c_dev_s *i2c, FAR
...
 		 */
-
 		priv->dev.ops = &g_audioops;
 		priv->lower = lower;
 		priv->i2c = i2c;
 		priv->i2s = i2s;
 
 		sem_init(&priv->devsem, 0, 1);
 		dq_init(&priv->pendq);
-		dq_init(&priv->doneq);
```
Comment (Taejun-Kwon) [inline os/drivers/audio/alc5658.c]: if we don't need doneq, why don't you remove doneq from structure in alc5658.h?(Actually I think we don't need it)

File: os/drivers/audio/alc5658.h [hunk 8]
```diff
@@ -109,7 +109,6 @@ struct alc5658_dev_s {
 	FAR struct i2c_dev_s *i2c;	/* I2C driver to use */
 	FAR struct i2s_dev_s *i2s;	/* I2S driver to use */
 	struct dq_queue_s pendq;	/* Queue of pending buffers to be sent */
-	struct dq_queue_s doneq;	/* Queue of sent buffers to be returned */
```
Comment (Taejun-Kwon) [inline os/drivers/audio/alc5658.h]: oops you already removed. ignore previous comment

File: os/drivers/audio/alc5658.c [hunk 9]
```diff
@@ -949,23 +959,14 @@ static int alc5658_ioctl(FAR struct audio_lowerhalf_s *dev, int cmd, unsigned lo
...
 
 	case AUDIOIOC_HWRESET: {
-		/* REVISIT:  Should we completely re-initialize the chip?   We
-		 * can't just issue a software reset; that would puts all ALC5658
-		 * registers back in their default state.
-		 */
-
-		audvdbg("AUDIOIOC_HWRESET:\n");
+		/* This should put ALC5658 in default state, reconfiguration needed */
+		audvdbg("AUDIOIOC_HWRESET: No Action Would be taken now \n");
```
Comment (Taejun-Kwon) [inline os/drivers/audio/alc5658.c]: We have a alc5658_hw_reset(), but we don't have to use here?

File: os/drivers/audio/alc5658.c [hunk 10]
```diff
@@ -674,8 +674,8 @@ static int alc5658_configure(FAR struct audio_lowerhalf_s *dev, FAR const struct
 	}
 
 	alc5658_setregs(priv);
-
 	alc5658_getregs(priv);
```
Comment (jsdosa) [inline os/drivers/audio/alc5658.c]: why do you setresgs and then getregs ?
Is it the code that checks to see if it is set?

File: os/drivers/audio/alc5658.c (hunk 9 above)
Comment (jsdosa) [reply os/drivers/audio/alc5658.c]: I think it is better to keep it.
Currently unused, it can be used according to the driver.

File: os/drivers/audio/alc5658.c (hunk 1 above)
Comment (Taejun-Kwon) [reply os/drivers/audio/alc5658.c]: After merge this & kishore's code #1157, I will update code regarding this.

File: os/drivers/audio/alc5658.c (hunk 2 above)
Comment (btheosam) [reply os/drivers/audio/alc5658.c]: return value of enqueue request would address timeout error.

File: os/drivers/audio/alc5658.c (hunk 6 above)
Comment (btheosam) [reply os/drivers/audio/alc5658.c]: But unknown cmd is a error.

File: os/drivers/audio/alc5658.c (hunk 7 above)
Comment (btheosam) [reply os/drivers/audio/alc5658.c]: Already removed in .h file also as part of this commit itself.

File: os/drivers/audio/alc5658.c (hunk 10 above)
Comment (btheosam) [reply os/drivers/audio/alc5658.c]: alc5658_getregs can be used to get the register values anytime for debugging alc5658.

File: os/drivers/audio/alc5658.c (hunk 1 above)
Comment (btheosam) [reply os/drivers/audio/alc5658.c]: Right now, the error flags are defined in i2s.h. flags are populated with the values defined in i2s.h. But based on tinyalsa changes & changes in audio layer, this can be modified.

File: os/drivers/audio/alc5658.c (hunk 2 above)
Comment (kishore-sn) [reply os/drivers/audio/alc5658.c]: No. Its not handled by enqueue request. 
Timeout happens much later after enqueue. It is indicated by result parameter of alc5658_rxtxcallback. But its being ignored.

Comment (kishore-sn) [reply os/drivers/audio/alc5658.c]: I think its better to just print an error message in alc5658_rxtxcallback when timeout error happens.

Comment (btheosam) [reply os/drivers/audio/alc5658.c]: Agreed

File: os/drivers/audio/alc5658.c (hunk 3 above)
Comment (btheosam) [reply os/drivers/audio/alc5658.c]: agreed

File: os/drivers/audio/alc5658.c (hunk 6 above)
Comment (Taejun-Kwon) [reply os/drivers/audio/alc5658.c]: No, it is not error case, it just undefined case for specific file operation(in this case audio operation of alc5658)
Any arg value can be sent to driver layer, so tizenrt usually using verbose for default value.

File: os/drivers/audio/alc5658.c (hunk 4 above)
Comment (btheosam) [reply os/drivers/audio/alc5658.c]: Just the order is changed. It is needed for setting sample rate.

**Comment by seinfra** [comment]
Code Rule Check OK.
